Added 22.1 docs on GC Pub/Sub sink

kathancox opened the pull request:
Fixes [DOC-1194](https://cockroachlabs.atlassian.net/browse/DOC-1194) [DOC-2228](https://cockroachlabs.atlassian.net/browse/DOC-1194)

Added docs for Google Cloud Pub/Sub sink:

- Description to Changefeed Sinks
- Example to Changefeed Examples
- Call outs for options 

Note: Added `Pub/Sub Editor` role as requirement for now, but I think this can be changed in a minor release with [Aaron's PR](https://github.com/cockroachdb/cockroach/pull/80430)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/13727)
<!-- Reviewable:end -->

kathancox commented:
Hey @amruss & @HonoreDB, as you review could you take a particular look at the options/parameters & known limitations? I added/edited to these but wasn't 100%, so would appreciate the double-check!

HonoreDB reviewed (approved):
Looks good to me! Added a couple of suggestions for additional info.

> Reviewed 2 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13727#-:-N0bd0aTEt_FninKze5V:bvg88dd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss and @kathancox)

___
*[`v22.1/changefeed-sinks.md` line 111 at r2](https://reviewable.io/reviews/cockroachdb/docs/13727#-N0baTrLBiVXXIDkEHM2:-N0baTrM21K_oVA6WKME:b-59gi1k) ([raw file](https://github.com/cockroachdb/docs/blob/c67d651c3bf47a6a62340df59fbd24f17f4436a2/v22.1/changefeed-sinks.md#L111)):*
> ```Markdown
> 
> - It only supports `JSON` message format.
> - Your Google Service Account must have the [Pub/Sub Editor](https://cloud.google.com/iam/docs/understanding-roles#pub-sub-roles) role.
> ```

And I think it has to have the role assigned *on the project*, not just on the topic you've pre-created. (I don't know a lot about gcloud permissions but it seemed like that was a distinction it was making).
___
*[`v22.1/changefeed-sinks.md` line 113 at r2](https://reviewable.io/reviews/cockroachdb/docs/13727#-N0bc2P278rFrGwyBPYl:-N0bc2P278rFrGwyBPYm:b-p0dbnf) ([raw file](https://github.com/cockroachdb/docs/blob/c67d651c3bf47a6a62340df59fbd24f17f4436a2/v22.1/changefeed-sinks.md#L113)):*
> ```Markdown
> - Your Google Service Account must have the [Pub/Sub Editor](https://cloud.google.com/iam/docs/understanding-roles#pub-sub-roles) role.
> - Changefeeds connecting to a Pub/Sub sink do not support the `topic_prefix` option.
> ```

Also, the `region` parameter is currently required. While GCP supports multi-region pubsub, they don't guarantee ordering across regions, so we need a single region to pin all output to.


<!-- Sent from Reviewable.io -->

kathancox commented:
> Looks good to me! Added a couple of suggestions for additional info.
> 
> > Reviewed 2 of 6 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13727#-:-N0bd0aTEt_FninKze5V:bvg88dd)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss and @kathancox)
> 
> _[`v22.1/changefeed-sinks.md` line 111 at r2](https://reviewable.io/reviews/cockroachdb/docs/13727#-N0baTrLBiVXXIDkEHM2:-N0baTrM21K_oVA6WKME:b-59gi1k) ([raw file](https://github.com/cockroachdb/docs/blob/c67d651c3bf47a6a62340df59fbd24f17f4436a2/v22.1/changefeed-sinks.md#L111)):_
> 
> > ```gfm
> > - It only supports `JSON` message format.
> > - Your Google Service Account must have the [Pub/Sub Editor](https://cloud.google.com/iam/docs/understanding-roles#pub-sub-roles) role.
> > ```
> 
> And I think it has to have the role assigned _on the project_, not just on the topic you've pre-created. (I don't know a lot about gcloud permissions but it seemed like that was a distinction it was making).
> 
> _[`v22.1/changefeed-sinks.md` line 113 at r2](https://reviewable.io/reviews/cockroachdb/docs/13727#-N0bc2P278rFrGwyBPYl:-N0bc2P278rFrGwyBPYm:b-p0dbnf) ([raw file](https://github.com/cockroachdb/docs/blob/c67d651c3bf47a6a62340df59fbd24f17f4436a2/v22.1/changefeed-sinks.md#L113)):_
> 
> > ```gfm
> > - Your Google Service Account must have the [Pub/Sub Editor](https://cloud.google.com/iam/docs/understanding-roles#pub-sub-roles) role.
> > - Changefeeds connecting to a Pub/Sub sink do not support the `topic_prefix` option.
> > ```
> 
> Also, the `region` parameter is currently required. While GCP supports multi-region pubsub, they don't guarantee ordering across regions, so we need a single region to pin all output to.

That's a good point Aaron... I should mention that this at the project level here too. 

Ack on the region, I'll add that in. Thanks so much!

kathancox commented:
Added Aaron's feedback & adjusted the descriptions of the URI params into a table

amruss reviewed (approved):
A couple minor nits but besides that LGTM

HonoreDB commented on v22.1/create-changefeed.md:
```diff
@@ -121,7 +139,7 @@ Option | Value | Description
 `webhook_client_timeout` | [`INTERVAL`](interval.html) |  If a response is not recorded from the sink within this timeframe, it will error and retry to connect. Note this must be a positive value. <br><br>**Default:** `"3s"`
 `webhook_auth_header` | [`STRING`](string.html) |  Pass a value (password, token etc.) to the HTTP [Authorization header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) with a webhook request for a "Basic" HTTP authentication scheme. <br><br> Example: With a username of "user" and password of "pwd", add a colon between "user:pwd" and then base64 encode, which results in  ...
-`topic_in_value` | [`BOOL`](bool.html) |  Set to include the topic in each emitted row update. Note this is automatically set for [webhook sinks](changefeed-sinks.html#webhook-sink).
+`topic_in_value` | [`BOOL`](bool.html) |  Set to include the topic in each emitted row update. Note this is automatically set for [webhook sinks](changefeed-sinks.html#webhook-sink) and [GC Pub/Sub sinks](changefeed-sinks.html#google-cloud-pub-sub).
```
Oh, good catch, no, this is not true for pubsub.

kathancox replied:
Ahh great, thanks for the confirmation, I will update that!

kathancox commented on v22.1/changefeed-sinks.md:
```diff
@@ -86,6 +87,42 @@ The configurable fields include:
+`AUTH`             | The authentication parameter can define either `specified` or `implicit` authentication. Passing your [Service Account](https://cloud.google.com/iam/docs/understanding-service-accounts) credentials with the URI or implicitly in your environment respectively. [Use Cloud Storage for Bulk Operations](use-cloud-storage-for-bulk-operations.html#authentication) provides more detail ...
+`CREDENTIALS`      | (Optional) The base64-encoded credentials of your Google [Service Account](https://cloud.google.com/iam/docs/understanding-service-accounts) credentials. Include with `AUTH=specified`.
+
+When using Pub/Sub as your downstream sink, consider the following:
```
I can add that you must specify a region. Instead of adding here that we are *planning* to add an unspecified/unordered option, what do you think about framing that within a limitation with a GH issue? We try not to add future plans in a "We're planning to add this in a future release..." to avoid either over-promising, or just general confusion on _when_ that could happen. 

Limitation could be framed as "X is not available in Pub/Sub [Tracking Issue]", which at least provides the user with context. (I can link to the limitation from this point as well.

kathancox commented on v22.1/changefeed-sinks.md:
```diff
@@ -86,6 +87,42 @@ The configurable fields include:
+`project name`     | The [Google Cloud Project](https://cloud.google.com/resource-manager/docs/creating-managing-projects) name.
+`region`           | The single region to which all output will be sent.
+`topic_name`       | (Optional) The topic name to which messages will be sent. See the following section on [Topic Naming](#topic-naming) for detail on how topics are created.
+`AUTH`             | The authentication parameter can define either `specified` or `implicit` authentication. Passing your [Service Account](https://cloud.google.com/iam/docs/understanding-service-accounts) credentials with the URI or implicitly in your environment respectively. [Use Cloud Storage for Bulk Operations](use-cloud-storage-for-bulk-operations.html#authentication) provides more detail ...
```
Yes, I shall add!

kathancox commented:
Hey @amruss @HonoreDB, I've made those changes and I'm going to pass on to docs review, but feel free to add any further comments on that!

kathancox commented:
Hey @andf-crl, thanks so much for the v helpful review, as per usual!

I incorporated all the changes except for a couple:

1. With the topic naming: I think that docs issue is particularly old. I tested with and without the `topic_name` param and it works fine 👍 
2. I have changed the language to "not supported" (much better, thanks). I dropped "Currently" though as we want to avoid future-looking language and just try to represent the feature-set as it stands. I added the known limitation so that we're showing that there is an improvement on a backlog there.
